packages/runtime/src/cache.tsx

```diff
@@ -0,0 +1,74 @@
+import { StyleSheetOpts } from './types';
+import { Bucket, buckets, getCompiledAttr } from './buckets';
+import { appendCSSTextNode, createStyleElement } from './css-utils';
+
+export const bucketsCache: Partial<{ [bucket in Bucket]: HTMLStyleElement }> = {};
+const bucketsAddedToHead: Partial<{ [bucket in Bucket]: boolean }> = {};
+let isBucketsCacheFilled = false;
+
+export function addBucketToHead(bucket: Bucket) {
+  if (!bucketsAddedToHead[bucket]) {
+    const bucketIndex = buckets.indexOf(bucket);
+    let nextBucketFromCache = null;
+
+    // Find next bucket before which we will add our current bucket element
+    for (let i = bucketIndex + 1; i < buckets.length; i++) {
+      nextBucketFromCache = bucketsCache[buckets[i]] as HTMLStyleElement;
+
+      if (document.head.contains(nextBucketFromCache)) {
+        break;
+      }
+
+      nextBucketFromCache = null;
+    }
+
+    document.head.insertBefore(bucketsCache[bucket] as HTMLStyleElement, nextBucketFromCache);
+
+    bucketsAddedToHead[bucket] = true;
+  }
+}
+
+function traverseStyleTextNode(
+  styleNode: HTMLStyleElement,
+  styleElement: HTMLStyleElement,
+  inserted: Record<string, true>
+) {
+  let styleNodeChild = styleNode.firstChild;
+
+  while (styleNodeChild) {
+    const css = styleNodeChild.textContent?.trim();
+
+    if (css) {
+      inserted[css] = true;
```

Comment:
We can also mark `inserted` `true` by checking if sheet is in `document.querySelector(style[data-compiled-${bucket}]` when its being rendered (on client) but that will become a performance bottleneck. This will run only once. Not sure how much we can rely on text nodes here.